src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1998 @@
+           of the array represented by that Reference.
+        3. A Binary Operation, in which case the code calls
+          `_handle_index_binop` to evaluate any additional dependencies.
+
+        Once these have been computed, any new dependencies are added into the
+        out_list, and the array reference itself will be added to the
+        shared_list if not already present.
+
+        :param ref: The Reference to be evaluated.
+        :type ref: :py:class:`psyclone.psyir.nodes.Reference`
+        :param private_list: The list of private References used in this task
+                             region.
+        :type private_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param firstprivate_list: The list of firstprivate References used in
+                                  this task region.
+        :type firstprivate_list: List of
+                                 :py:class:`psyclone.psyir.nodes.Reference`
+        :param shared_list: The list of shared References for this task.
+        :type shared_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+        :param out_list: The list of output References for this task.
+        :type out_list: List of :py:class:`psyclone.psyir.nodes.Reference`
+
+        :raises GenerationError: If an array index is a shared variable.
+        """
+        # We write to this arrayref, so its shared and depend out on
+        # the array.
+
+        # Arrays are always shared at the moment, so we ignore the possibility
+        # of it being private now.
+
+        # Index list stores the set of indices to use with this ArrayMixin
+        # for the depend clause.
+        index_list = []
+        # Work out the indices needed.
+        for dim, index in enumerate(ref.indices):
+            if isinstance(index, Literal):
+                # Literals are just a value, just use the value.
+                index_list.append(index.copy())
+            elif isinstance(index, Reference):
+                index_private = self._is_reference_private(index)
+                # Check whether the reference is to a child loop variable.
+                child_loop_vars = []
+                for child_loop in self.walk(Loop):
+                    if child_loop.variable not in self._proxy_loop_vars:
+                        child_loop_vars.append(child_loop.variable)
+
+                if index_private:
+                    if (
+                        index not in private_list
+                        and index not in firstprivate_list
+                    ):
+                        firstprivate_list.append(index.copy())
+                    # Special case 1. If index belongs to a child loop
+                    # that is NOT a proxy for a parent loop, then we
+                    # can only do as well as guessing the entire range
+                    # of the loop is used.
+                    if index.symbol in child_loop_vars:
```

Comment:
There's an awful lot of duplication here wrt some of the earlier methods.

Reply:
@arporter Do you want me to refactor these repeats out to "_gen_full_range" or similar once they're using the `ArrayMixin` calls or shall I just leave the repeated code?
At best I think we can get these down to:
`index_list.append(self._gen_full_range(...))` - this is then quite self explanatory so might be ok?

Reply:
This is done.

Reply:
That's better, thanks.

Reply:
See earlier discussion on duplication.